src/protobuf.h

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright The async-profiler authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _PROTOBUF_H
+#define _PROTOBUF_H
+
+#include "arch.h"
+#include <string.h>
+#include <type_traits>
+
+typedef const u32 protobuf_t;
+static protobuf_t VARINT = 0;
+static protobuf_t LEN = 2;
+
+typedef u32 protobuf_index_t;
+
+// We assume the length of a nested field can be represented with 3 varint
+// bytes.
+const size_t nested_field_byte_count = 3;
+
+class ProtobufBuffer {
+private:
+  ProtobufBuffer *_parent_message;
+  char *_data;
+  size_t _offset;
+
+  ProtobufBuffer(ProtobufBuffer *parent) :
+    _data(parent->_data + parent->offset()),
+    _offset(0),
+    _parent_message(parent) {}
+
+  template <typename T>
+  typename std::enable_if<std::is_unsigned<T>::value, void>::type
+  putVarInt(T n);
+
+  template <typename T>
+  typename std::enable_if<std::is_unsigned<T>::value, size_t>::type
+  putVarInt(size_t offset, T n);
+
+  void tag(protobuf_index_t index, protobuf_t type);
+
+  void commitMessage(size_t message_length);
+
+public:
+  ProtobufBuffer(char *data) :
+    _data(data),
+    _offset(0),
+    _parent_message(nullptr) {}
+  ~ProtobufBuffer();
+
+  const char *data() const { return _data; }
+
+  size_t offset() const { return _offset; }
+
+  // VARINT
+  void field(protobuf_index_t index, bool b);
+  void field(protobuf_index_t index, u32 n);
+  void field(protobuf_index_t index, u64 n);
+  // LEN
+  void field(protobuf_index_t index, const char *s);
+  void field(protobuf_index_t index, const char *s, size_t len);
+  void field(protobuf_index_t index, const ProtobufBuffer buffer, size_t len);
+
+  ProtobufBuffer startMessage(protobuf_index_t index);
+};
+
+template <typename T>
+typename std::enable_if<std::is_unsigned<T>::value, void>::type
+ProtobufBuffer::putVarInt(T n) {
+  _offset = putVarInt(_offset, n);
+}
+
+template <typename T>
+typename std::enable_if<std::is_unsigned<T>::value, size_t>::type
+ProtobufBuffer::putVarInt(size_t offset, T n) {
+  while ((n >> 7) != 0) {
+    _data[offset++] = (char)(0b10000000 | (n & 0b01111111));
```

Comment:
No binary literals, please. My eyes are bleeding when counting ones and zeros.
Every system developer should know `0x80` and `0x7f`.

Reply:
d5644f8deed2e5c350ae62f74c32c07fafacaf05